terraform/autoneg/main.tf

```diff
@@ -0,0 +1,335 @@
+    verbs      = ["create", "patch"]
+  }
+
+  rule {
+    api_groups = [""]
+    resources  = ["services"]
+    verbs      = ["get", "list", "patch", "update", "watch"]
+  }
+
+  rule {
+    api_groups = [""]
+    resources  = ["services/status"]
+    verbs      = ["get", "patch", "update"]
+  }
+}
+
+resource "kubernetes_cluster_role" "clusterrole_autoneg_proxy_role" {
+  metadata {
+    name = "autoneg-proxy-role"
+
+    labels = {
+      app = "autoneg"
+    }
+    annotations = {}
+  }
+
+  rule {
+    api_groups = ["authentication.k8s.io"]
+    resources  = ["tokenreviews"]
+    verbs      = ["create"]
+  }
+
+  rule {
+    api_groups = ["authorization.k8s.io"]
+    resources  = ["subjectaccessreviews"]
+    verbs      = ["create"]
+  }
+}
+
+resource "kubernetes_role_binding" "rolebinding_autoneg_leader_election_rolebinding" {
+  metadata {
+    namespace = kubernetes_namespace.namespace_autoneg_system.metadata[0].name
+    name      = "autoneg-leader-election-rolebinding"
+
+    labels = {
+      app = "autoneg"
+    }
+    annotations = {}
+  }
+
+  role_ref {
+    api_group = "rbac.authorization.k8s.io"
+    kind      = "Role"
+    name      = kubernetes_role.role_autoneg_leader_election_role.metadata[0].name
+  }
+  subject {
+    kind      = "ServiceAccount"
+    name      = kubernetes_service_account.service_account.metadata[0].name
+    namespace = kubernetes_namespace.namespace_autoneg_system.metadata[0].name
+  }
+}
+
+resource "kubernetes_cluster_role_binding" "clusterrolebinding_autoneg_manager_rolebinding" {
+  metadata {
+    name = "autoneg-manager-rolebinding"
+    labels = {
+      app = "autoneg"
+    }
+    annotations = {}
+  }
+  role_ref {
+    api_group = "rbac.authorization.k8s.io"
+    kind      = "ClusterRole"
+    name      = kubernetes_cluster_role.clusterrole_autoneg_manager_role.metadata[0].name
+  }
+  subject {
+    kind      = "ServiceAccount"
+    name      = kubernetes_service_account.service_account.metadata[0].name
+    namespace = kubernetes_namespace.namespace_autoneg_system.metadata[0].name
+  }
+}
+
+resource "kubernetes_cluster_role_binding" "clusterrolebinding_autoneg_proxy_rolebinding" {
+  metadata {
+    name = "autoneg-proxy-rolebinding"
+    labels = {
+      app = "autoneg"
+    }
+    annotations = {}
+  }
+  role_ref {
+    api_group = "rbac.authorization.k8s.io"
+    kind      = "ClusterRole"
+    name      = kubernetes_cluster_role.clusterrole_autoneg_proxy_role.metadata[0].name
+  }
+  subject {
+    kind      = "ServiceAccount"
+    name      = kubernetes_service_account.service_account.metadata[0].name
+    namespace = kubernetes_namespace.namespace_autoneg_system.metadata[0].name
+  }
+}
+
+resource "kubernetes_service" "service_autoneg_controller_manager_metrics_service" {
+  metadata {
+    annotations = {
+      "prometheus.io/port"   = "8443"
+      "prometheus.io/scheme" = "https"
+      "prometheus.io/scrape" = "true"
+      "cloud.google.com"     = "'{\"ingress\":true}'"
```

Comment:
I added this because GKE seems to add it on apply and next run would result it getting removed. I think for it to be externally exposed, you'd have to have a matching Ingress resource.

```tf
  # module.autoneg.module.kubernetes.kubernetes_service.service_autoneg_controller_manager_metrics_service will be updated in-place
  ~ resource "kubernetes_service" "service_autoneg_controller_manager_metrics_service" {
        id                     = "autoneg-system/autoneg-controller-manager-metrics-service"
        # (2 unchanged attributes hidden)

      ~ metadata {
          ~ annotations      = {
              - "cloud.google.com/neg" = jsonencode(
                    {
                      - ingress = true
                    }
                ) -> null
                # (3 unchanged elements hidden)
            }
            name             = "autoneg-controller-manager-metrics-service"
            # (5 unchanged attributes hidden)
        }

        # (1 unchanged block hidden)
    }
```

Reply:
You are correct. This (poorly named?) annotation enables the NEG creation for the service. 

https://cloud.google.com/kubernetes-engine/docs/concepts/ingress#container-native_load_balancing

Reply:
I will leave it in - shouldn't do any harm.

Reply:
It’s unlikely that the metrics service NEGs will be added behind a load balancer since metric collection is done directly against each instance explicitly. Is there a NEG quota that these would count against that could become a problem in projects with many services in them?

The terraform reapply does need to be addressed. Is it possible to add a lifecycle ignore on that annotation entry so terraform won’t try to update it?

Reply:
I don't think there is a quota. I do think it's better if it's stable after apply, I will just set it to an empty notation (`{}`). That seems to work.

Reply:
For future readers, there is a per project NEG quota that currently defaults to 100, which can be requested to be increased well beyond that default. This would just add one more NEG per cluster of set to true so quota shouldn’t be a concern.